libflux/src/semantic/walk/walk_mut.rs

```diff
@@ -142,6 +144,125 @@ impl<'a> NodeMut<'a> {
             NodeMut::MemberAssgn(n) => &n.loc,
         }
     }
+    pub fn type_of(&self) -> Option<&MonoType> {
+        match self {
+            NodeMut::IdentifierExpr(n) => Some(&n.typ),
+            NodeMut::ArrayExpr(n) => Some(&n.typ),
+            NodeMut::FunctionExpr(n) => Some(&n.typ),
+            NodeMut::LogicalExpr(n) => Some(&n.typ),
+            NodeMut::ObjectExpr(n) => Some(&n.typ),
+            NodeMut::MemberExpr(n) => Some(&n.typ),
+            NodeMut::IndexExpr(n) => Some(&n.typ),
+            NodeMut::BinaryExpr(n) => Some(&n.typ),
+            NodeMut::UnaryExpr(n) => Some(&n.typ),
+            NodeMut::CallExpr(n) => Some(&n.typ),
+            NodeMut::ConditionalExpr(n) => Some(&n.typ),
+            NodeMut::StringExpr(n) => Some(&n.typ),
+            NodeMut::IntegerLit(n) => Some(&n.typ),
+            NodeMut::FloatLit(n) => Some(&n.typ),
+            NodeMut::StringLit(n) => Some(&n.typ),
+            NodeMut::DurationLit(n) => Some(&n.typ),
+            NodeMut::UintLit(n) => Some(&n.typ),
+            NodeMut::BooleanLit(n) => Some(&n.typ),
+            NodeMut::DateTimeLit(n) => Some(&n.typ),
+            NodeMut::RegexpLit(n) => Some(&n.typ),
+            _ => None,
+        }
+    }
+    pub fn replace_type(&mut self, sub: &Substitution) {
+        match self {
+            NodeMut::IdentifierExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::ArrayExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::FunctionExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::LogicalExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::ObjectExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::MemberExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::IndexExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::BinaryExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::UnaryExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::CallExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::ConditionalExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::StringExpr(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::IntegerLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::FloatLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::StringLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::DurationLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::UintLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::BooleanLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::DateTimeLit(ref mut n) => n.typ = n.typ.apply(sub),
+            NodeMut::RegexpLit(ref mut n) => n.typ = n.typ.apply(sub),
+            _ => (),
+        }
+    }
```

Comment:
Ok I think I understand. Is this why you changed `Substitutable`s to take `&self` instead `self`? Because otherwise you'd have to clone here?

But `Node` and `NodeMut` are public types right? Visible from external crates? By implementing `replace_type` on `Node` won't we then have to make the `Substitution` type a public item of this crate? I don't think we want that. Substitutions should only be used internally during type inference.

Perhaps a `visitor` isn't the right job for this. Perhaps `Expression` types should just be `Substitutable`? Something like this?

```Rust
impl Substitutable for Expression {
    fn apply(mut self, sub: &Substitution) -> Self {
        match self {
            Expression::Integer(mut lit) => {
                let typ = lit.typ.apply(&sub);
                lit.typ = typ;
                Expression::Integer(lit)
            }
            ...
            Expression::Binary(mut exp) => {
                let typ = exp.typ.apply(&sub);
                exp.typ = typ;
                Expression::Binary(exp)
            }
            ...
        }
    }
    fn free_vars(&self) -> Vec<Tvar> {
        self.type_of().free_vars()
    }
}
```

Reply:
About visibility, I have doubts about this, because `cargo` doesn't warn you or even error if you expose as public something that's "partially" public. So, I don't exactly know what the outcome of invoking `replace_type` without possibly passing a `Substitution` would be. I'll check.

Anyways, your strategy seems pretty elegant. I could `apply` to the expression and replace the reference to the Node in `NodeMut`. This seems to be a great desing, because the Expression itself should be able to move its pieces out to build the return value, thus avoiding cloning! I will try it out today. I am still doubtful about partial moves and lifetimes for pointers in `NodeMut`, but I'll try.

Thanks for highlighting this alternative approach Joshua 😎